### PR TITLE
[experimental pr] Support more cases when analyzing fence space

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 1]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
         "__fetch_builtin_z",
-        "uint4"};
+        "uint4",
+        "sqrtf",
+        "__expf"};
+
+bool clang::dpct::GlobalPointerReferenceCountAnalyzer::Visit(
+    clang::DeclRefExpr *DRE) {
+  // Collect all DREs and its Decl
+  clang::VarDecl *VarD = dyn_cast_or_null<VarDecl>(DRE->getDecl());
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: move to 375

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 2]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+  clang::FieldDecl *FieldD = dyn_cast_or_null<FieldDecl>(DRE->getDecl());
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: move to 377

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 3]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+  clang::ValueDecl *VD = nullptr;
+  if (VarD) {
+    VD = VarD;
+  } else if (FieldD) {
+    VD = FieldD;
+  } else {
+    return true;
+  }
+  if (VD->hasAttr<clang::CUDADeviceAttr>() && isUserDefinedDecl(VD)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: add managed attr

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 4]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+      }
+    }
+  }
+}
+
+bool clang::dpct::GlobalPointerReferenceCountAnalyzer::countReference(
+    const clang::FunctionDecl *FD) {
+  // Collect all non-const pointers which point to fundamental type
+  std::unordered_set<clang::VarDecl *> NonconstPointerDecls;
+  if (!FD->hasAttr<clang::CUDAGlobalAttr>())
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: remove

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 5]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+    return false;
+  if (HasGlobalDeviceVariable)
+    return false;
+  for (auto &Param : FD->parameters()) {
+    if (!Param->getType()->isPointerType() &&
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: merge line 464 and 468

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 6]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+  } else {
+    I->second.insert(DRE);
+  }
+  return true;
+}
+void clang::dpct::GlobalPointerReferenceCountAnalyzer::PostVisit(
+    clang::DeclRefExpr *) {}
+
+namespace {
+const clang::VarDecl *isAssignAlias(clang::DeclRefExpr *DRE) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: change the function name

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 7]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+    return VD;
+  }
+  const clang::BinaryOperator *BO =
+      clang::dpct::DpctGlobalInfo::findAncestor<clang::BinaryOperator>(DRE);
+  if (!BO)
+    return nullptr;
+
+  if (auto CE = dyn_cast<clang::CallExpr>(BO->getRHS())) {
+    if (auto FD = CE->getDirectCallee()) {
+      if (!isUserDefinedDecl(FD)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: if is cuda, check allow list

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 8]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+                                NewNonconstPointerDecls.end());
+  } while (NonconstPointerDeclsSize != NonconstPointerDecls.size());
+  // Check if pointers in NonconstPointerDecls is passed to other functions
+  for (const auto VD : NonconstPointerDecls) {
+    auto I = DeclDREsMap.find(VD);
+    if (I == DeclDREsMap.end())
+      continue;
+    for (auto const &DRE : I->second) {
+      if (auto FuncCall =
+              clang::dpct::DpctGlobalInfo::findAncestor<clang::CallExpr>(DRE)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: add a table to find

File: clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp [hunk 9]
```diff
@@ -327,14 +352,198 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
...
+    }
+  }
+
+  // If pointers in NonconstPointerDecls only used once, then we can use
+  // local_space
+  for (const auto VD : NonconstPointerDecls) {
+    auto I = DeclDREsMap.find(VD);
+    if (I == DeclDREsMap.end())
+      continue;
+    if (I->second.size() > 1) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp]: need merge DRE usages in alias

File: clang/test/dpct/syncthreads.cu [hunk 10]
```diff
@@ -149,4 +161,116 @@ __global__ void test9(S4 a) {
   S4 b(a);
   // CHECK:item_ct1.barrier();
   __syncthreads();
-}
\ No newline at end of file
+}
+
+extern __shared__ float extern_local_decl[];
+
+__global__ void test10(float *arg1, int arg2, int arg3) {
```
Comment (zhimingwang36) [inline clang/test/dpct/syncthreads.cu]: arg1 ==> arg1_ptr;   arg2==> arg2_scalar ;  arg3 ==> arg3_scalar

Comment (zhiweij1) [reply clang/test/dpct/syncthreads.cu]: done

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 11]
```diff
@@ -724,11 +732,12 @@ void DpctFileInfo::buildUnionFindSetForUncalledFunc() {
 }
 
 void DpctFileInfo::buildKernelInfo() {
-  for (auto &Kernel : KernelMap)
+  for (auto &Kernel : KernelMap) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: undo this change

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 12]
```diff
@@ -942,15 +951,18 @@ void DpctFileInfo::buildReplacements() {
     addReplacement(R);
   }
 
-  for (auto &Kernel : KernelMap)
+  for (auto &Kernel : KernelMap) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: undo this change

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 13]
```diff
@@ -685,7 +685,7 @@ void ExprAnalysis::analyzeExpr(const MemberExpr *ME) {
   auto ItemItr = NdItemMap.find(BaseType);
   if (ItemItr != NdItemMap.end()) {
     if (MapNames::replaceName(NdItemMemberMap, FieldName)) {
-      if (DpctGlobalInfo::getAssumedNDRangeDim() == 1) {
+      {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]: remove this brace

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 14]
```diff
@@ -11977,18 +11949,9 @@ void SyncThreadsRule::runRule(const MatchFinder::MatchResult &Result) {
...
-      emplaceTransformation(new ReplaceStmt(CE, std::move(Replacement)));
-    } else {
-      report(CE->getBeginLoc(), Diagnostics::BARRIER_PERFORMANCE_TUNNING, true,
-             "nd_item");
-      std::string Replacement = DpctGlobalInfo::getItem(CE) + ".barrier()";
-      emplaceTransformation(new ReplaceStmt(CE, std::move(Replacement)));
-    }
+    // Here we only do analysis, replacement will be generated in the migration
+    // pass.
+    DpctGlobalInfo::getItem(CE);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp:11954]: for readability,  please add new interface for the ops done in getItem(), which target to be used here.

File: clang/test/dpct/syncthreads.cu [hunk 15]
```diff
@@ -149,4 +161,129 @@ __global__ void test9(S4 a) {
...
   // CHECK:item_ct1.barrier();
   __syncthreads();
-}
\ No newline at end of file
+}
+
+extern __shared__ float extern_local_decl[];
+
+__global__ void test10(float *arg1_ptr, int arg2_scalar, int arg3_scalar) {
+  int var1 = blockIdx.x * blockDim.x + threadIdx.x;
```
Comment (zhiweij1) [inline clang/test/dpct/syncthreads.cu:169]: we can check the item_ct1's type, migrate syncthread to an if-else statement.
Then we can get rid of the analysis of 1dim/3dim of kernel call.

File: clang/test/dpct/syncthreads.cu [hunk 16]
```diff
@@ -149,4 +161,129 @@ __global__ void test9(S4 a) {
...
+  if (var1 < arg3_scalar) {
```
Comment (zhiweij1) [inline clang/test/dpct/syncthreads.cu:170]: also, we cannot determine syncthread call one by one.
We need to analysis in an iteration, because previous change (global => local) will affect currect analysis.

File: clang/test/dpct/syncthreads.cu (hunk 15 above)
Comment (zhiweij1) [reply clang/test/dpct/syncthreads.cu:169]: If we keep this 1dim/3dim analysis, we need to add more tests.
E.g.,
two different kernel calls with same global function;
no kernel call with this global function.
